### PR TITLE
feat(forge): allow invariant tests to target interfaces

### PR DESCRIPTION
**Pull request by bernard-wagner**
## Motivation

Currently it pretty difficult to perform fuzzing of contracts that make use of diamond proxies, other routing proxies or even more broadly any existing contract in a forked environment. Even when overloading `targetSelectors` and `targetContracts` only contracts kept in `known_contracts` are valid targets. 

## Solution

Add a new invariant helper function that allows associating a contract's or interface's Abi to an address for fuzzing purposes. Effectively making the address `known` without having to deploy the contract during `setUp`.

### CODE REVIEW

File: evm/src/fuzz/invariant/executor.rs [hunk 1]
```diff
@@ -407,6 +407,25 @@ impl<'a> InvariantExecutor<'a> {
...
+            {
+                // Merge the functions of multiple contracts with the same address
+                combined
+                    .entry(*key)
+                    .or_insert_with(|| (identifier.clone(), abi.clone(), vec![]))
+                    .1
+                    .functions
+                    .extend(abi.functions.clone());
+            }
+        }
```
Comment (mattsse) [inline evm/src/fuzz/invariant/executor.rs:427]: this looks right, would love to see some additional docs on what exactly is happening here

**Review by mattsse — CHANGES_REQUESTED** [review]
defer to @Evalir mostly, but sounds reasonable

only have doc nits

**Comment by mds1** [comment]
I agree there should be better support for targeting proxies, and agree with the general approach here, but I think instead of calling it `targetProxies` we should frame it more generically, since other use cases include targeting non-proxies where forge only has bytecode (for example, if forking mainnet and a contract isn't verified). Or, using a trimmed down interface to only target part of a contract without having to enumerate all selectors.

We currently have a `FuzzSelector` struct, and I think adding a `FuzzArtifact` struct which is similar to your `FuzzTarget` struct is the way to go::
```solidity
struct FuzzArtifact {
  address target;
  string[] artifacts;
}
```

This way you can pass arbitrary `(address, selector[])` and/or `(address, artifact[]`) pairings for a given address, regardless of whether or not something is a proxy

I'm not sure of a good name for the method though, maybe `targetArtifactMap` or something? Don't love that though

**Comment by bernard-wagner** [comment]
> I agree there should be better support for targeting proxies, and agree with the general approach here, but I think instead of calling it `targetProxies` we should frame it more generically, since other use cases include targeting non-proxies where forge only has bytecode (for example, if forking mainnet and a contract isn't verified). Or, using a trimmed down interface to only target part of a contract without having to enumerate all selectors.
> 
> We currently have a `FuzzSelector` struct, and I think adding a `FuzzArtifact` struct which is similar to your `FuzzTarget` struct is the way to go::
> 
> ```solidity
> struct FuzzArtifact {
>   address target;
>   string[] artifacts;
> }
> ```
> 
> This way you can pass arbitrary `(address, selector[])` and/or `(address, artifact[]`) pairings for a given address, regardless of whether or not something is a proxy
> 
> I'm not sure of a good name for the method though, maybe `targetArtifactMap` or something? Don't love that though

Think most of these points make a lot of sense.

> I agree there should be better support for targeting proxies, and agree with the general approach here, but I think instead of calling it `targetProxies` we should frame it more generically, since other use cases include targeting non-proxies where forge only has bytecode (for example, if forking mainnet and a contract isn't verified). Or, using a trimmed down interface to only target part of a contract without having to enumerate all selectors.
> 
> We currently have a `FuzzSelector` struct, and I think adding a `FuzzArtifact` struct which is similar to your `FuzzTarget` struct is the way to go::
> 
> ```solidity
> struct FuzzArtifact {
>   address target;
>   string[] artifacts;
> }
> ```
> 
> This way you can pass arbitrary `(address, selector[])` and/or `(address, artifact[]`) pairings for a given address, regardless of whether or not something is a proxy
> 
> I'm not sure of a good name for the method though, maybe `targetArtifactMap` or something? Don't love that though

Thanks for the feedback!

Not sure if Artifact conveys the use-case that well. Would the following naming convention be more suitable? 
```solidity
struct FuzzInterface {
  address target;
  string[] artifacts;
}

function targetInterfaces() public returns (FuzzInterface[] memory)
```

Especially given I am still investigating how to also allow specifying interfaces (artifacts with no bytecode), which would fit the desired use case pretty well.

File: evm/src/fuzz/invariant/executor.rs (hunk 1 above)
Comment (bernard-wagner) [reply evm/src/fuzz/invariant/executor.rs]: Think it a lot better now. Reworked to destructure the entry and avoid `.1.` and expanded the documentation.

**Comment by mds1** [comment]
> Not sure if Artifact conveys the use-case that well. Would the following naming convention be more suitable? 
> ```solidity
> struct FuzzInterface {
>   address target;
>   string[] artifacts;
> }
> 
> function targetInterfaces() public returns (FuzzInterface[] memory)
> ```

Ah nice, yea I like these names! cc @lucas-manuel too for thoughts

File: utils/src/lib.rs [hunk 2]
```diff
@@ -225,7 +225,22 @@ pub fn link_with_nonce_or_address<T, U>(
                 }
                 BytecodeObject::Bytecode(ref bytes) => {
                     if bytes.as_ref().is_empty() {
-                        // abstract, skip
+                        // Handle case where bytecode bytes are empty
```
Comment (bernard-wagner) [inline utils/src/lib.rs:228]: @mattsse This is how I am currently getting Interfaces to be included in the `known_contracts`. Might add a bit of over overhead as each Abstract contract and interface's abi will be kept in memory. Any objections to this approach?

File: cli/src/cmd/forge/script/build.rs [hunk 3]
```diff
@@ -167,8 +167,10 @@ impl ScriptArgs {
                     }
                 }
 
-                let tc: ContractBytecode = contract.into();
-                highlevel_known_contracts.insert(id, tc.unwrap());
+                if let Ok(tc) = ContractBytecode::from(contract).try_into() {
```
Comment (bernard-wagner) [inline cli/src/cmd/forge/script/build.rs:170]: Needed to not panic if `bytecode` is empty.

**Comment by Evalir** [comment]
hey @bernard-wagner — will review this soon, but we recently flattened all crates into `crates/` so there's some conflicts to solve, sorry!

**Comment by bernard-wagner** [comment]
@Evalir I rebased, so it should be a bit easier to review.

File: crates/cli/src/forge/cmd/script/build.rs [hunk 4]
```diff
@@ -167,8 +167,10 @@ impl ScriptArgs {
                     }
                 }
 
-                let tc: ContractBytecode = contract.into();
-                highlevel_known_contracts.insert(id, tc.unwrap());
+                if let Ok(tc) = ContractBytecode::from(contract).try_into() {
```
Comment (Evalir) [inline crates/cli/src/forge/cmd/script/build.rs:170]: why is this a `try_into` now?

File: crates/evm/src/fuzz/invariant/executor.rs [hunk 5]
```diff
@@ -428,6 +428,38 @@ impl<'a> InvariantExecutor<'a> {
...
+        let interfaces =
+            self.get_list::<(Address, Vec<String>)>(invariant_address, abi, "targetInterfaces");
+
+        // `combined` is a BTreeMap that will store contracts mapped by their target address.
+        // This map is used to merge functions of the specified interfaces for the same address.
+        let mut combined: TargetedContracts = BTreeMap::new();
+
+        // Loop through each address and its associated artifact identifiers.
+        // We're borrowing the interfaces here to avoid taking full ownership of them.
+        for (key, identifiers) in &interfaces {
```
Comment (Evalir) [inline crates/evm/src/fuzz/invariant/executor.rs:440]: All of this is done for being able to use interfaces/proxies right? Could we add a little bit of docs on how this helps achieve this usage? without context it's hard to understand that this is for diamong proxies

File: crates/forge/src/multi_runner.rs [hunk 6]
```diff
@@ -275,15 +275,21 @@ impl MultiContractRunnerBuilder {
...
+                        bytes
                     } else {
+                        known_contracts.insert(id.clone(), (abi, vec![]));
                         return Ok(())
-                    };
+                    }
+                } else {
+                    known_contracts.insert(id.clone(), (abi, vec![]));
+                    return Ok(())
+                };
```
Comment (Evalir) [inline crates/forge/src/multi_runner.rs:291]: Can we get the rationale for this change? This code is extremely sensible—it touches how we link contracts and we've had serious issues in the past on this.

**Review by Evalir — CHANGES_REQUESTED** [review]
I am not entirely against adding this—it'd be useful. However, this touches extremely sensible code paths on which we've performed several fixes recently so I'd like to see much better documentation, and more tests as well.

File: crates/cli/src/forge/cmd/script/build.rs (hunk 4 above)
Comment (bernard-wagner) [reply crates/cli/src/forge/cmd/script/build.rs:170]: `tc.unwrap()` panics if `bytecode` or `deployed_bytecode` is `None`, which is the case for `interfaces`. So the code checks whether the contract is deployable with `try_into()`. As far as I understand it `highlevel_known_contracts` should only contain deployable contracts.

**Comment by bernard-wagner** [comment]
@Evalir somehow I just totally missed `targetArtifacts` which already exists making the introduction of an additional helper function pretty defunct (feeling pretty dumb right now). Apologies. Will rework to just add the ability to associate `interfaces` to the existing `targetArtifacts` helper function. 

Change should just be to enable `targetArtifacts` to include ABIs that are not deployable.

File: crates/evm/src/fuzz/invariant/executor.rs (hunk 5 above)
Comment (bernard-wagner) [reply crates/evm/src/fuzz/invariant/executor.rs:440]: Effectively making the address known without having to deploy the contract during setUp. Will expand on the docs to include this info.

The issue is not really just diamond proxies but really just interface based fuzzing. I develop this I realized the use case is a lot more aligned for fork based fuzzing, or even fuzzing Vyper contracts. Diamond and Router proxies is just the use case I started with.

File: crates/forge/src/multi_runner.rs (hunk 6 above)
Comment (bernard-wagner) [reply crates/forge/src/multi_runner.rs:291]: As mentioned in another comment, currently any abi without bytecode is ignored. So if the contract is not deployable we just add it to `known_contracts`, but not `deployable_contracts`.

Ideally looking for the inverse of the `and_then` chain at the end of the function.

File: utils/src/lib.rs (hunk 2 above)
Comment (Evalir) [reply utils/src/lib.rs]: I think this is fine for now—we can profile and optimize later if it slows things down.

**Review by Evalir — APPROVED** [review]
right so i tested this a bit and seems good—lgtm, i suppose we need work on forge std too right? apart from the book docs as usual

**Pull request merged**